rfcs/20231012-external-mel-bindings.md

```diff
@@ -0,0 +1,546 @@
+get transformed into regular strings in externals.
+
+The syntactic construct chosen is also consistent with the generated code:
+unlike `external` declarations, `let` bindings don't get inlined at the call
+point.
+
+If the author of the binding doesn't want to use labeled arguments, they can
+append an underscore to have them removed from the generated OCaml function:
+
+```ocaml
+let%js decodeURI : uri_:string -> string = "decodeURI($uri_)"
+let _ : string = decodeURI "foo" (* no warning about a missing label ever *)
+```
+
+The preprocessor of `let%js` will check that type definitions to have at least
+an arrow type. So in cases where before one would just have some abstract type
+`t` in the external, now they would become `unit -> t`.
+
+For example, the current external to `Math.PI`:
+
+```ocaml
+external pi : float = "PI" [@@mel.scope "Math"]
+```
+
+would be written as:
+
+```ocaml
+let%js pi : float = "Math.PI"
+```
+
+## `let%js.import`
+
+The other extension introduced, `let%js.import`, is used for importing values
+from JavaScript modules into the current OCaml module:
+
+```ocaml
+type t
+let%js.import foo: t = "./bar"
+```
+
+Gets transformed into this JavaScript:
+
+```js
+import foo from "./bar";
+```
+
+It can leverage the OCaml destructuring feature to pick values from the imported
+module, but all of them need type annotations:
+
+
+```ocaml
+type u
+let%js.import {foo = (hey : string); bar: int} = "./bar"
+```
+Gets transformed into this JavaScript:
+
+```js
+import {foo as hey, bar} from "./bar";
+```
+
+Note there is no need to predefine a record type before using `let%js.import`,
+as the record is used merely syntactically to be transformed into an `import`
+statement.
+
+## `let%js.obj`
+
+Would be similar to `mel.obj` but as an extension applied to `let` bindings:
+
+```ocaml
+let%js.obj john = { name = "john"; age = 99 }
+```
+
+# Motivation
+
+The current design of the bindings to JavaScript code in Melange has a few
+shortcomings:
+
+- They are not particularly user friendly, as writing attributes like
+`mel.scope`, `mel.send`, `mel.module` is very different to what one would write
+in JavaScript.
+
+- They are not friendly to OCaml native compilation either, which makes writing
+share libraries harder. The friction exists because Melange bindings support
+externals with a simple type like `t`, while OCaml only supports externals with
+arrow types like `t -> u`. One can process the AST to fix this, but given that
+they are just plain `external` items, one can't leverage the existing
+context-free rule for extensions infrastructure in ppxlib, and has to process
+the full AST with a mapper.
+
+- Another source of friction comes from the optimizations done by the compiler
+in `external` statements. Melange compilation model is module-oriented, unlike
+js_of_ocaml, which has an executable-oriented compilation model. In Melange,
+users expect that the code referenced by an external function will be compiled
```

Comment:
is this an actual expectation of users?

Reply:
I think it is, either consciously or unconsciously, considering how confused we all are when bugs related to this happen.